src/main/java/com/blamejared/crafttweaker/impl/predicate/StatePropertiesPredicate.java

```diff
@@ -0,0 +1,160 @@
+package com.blamejared.crafttweaker.impl.predicate;
+
+import com.blamejared.crafttweaker.api.annotations.ZenRegister;
+import com.blamejared.crafttweaker_annotations.annotations.Document;
+import net.minecraft.state.Property;
+import net.minecraft.state.StateContainer;
+import net.minecraft.state.StateHolder;
+import org.openzen.zencode.java.ZenCodeType;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.BiPredicate;
+import java.util.stream.Collectors;
+
+@ZenRegister
+@ZenCodeType.Name("crafttweaker.api.predicate.StatePropertiesPredicate")
+@Document("vanilla/api/predicate/StatePropertiesPredicate")
+public final class StatePropertiesPredicate extends IVanillaWrappingPredicate.AnyDefaulting<net.minecraft.advancements.criterion.StatePropertiesPredicate> {
+    private static class RawMatcher {
```

Comment:
Split this into two Matcher classes, ranged and exact?
You could extract an interface if you need them all in the same map below.

Or you could do something like
```
Map<String, Either<ExactRawMatcher, RangedRawMatcher>>
```
Though a simple interface may be easier to read lol